ostd/src/cpu/mod.rs

```diff
@@ -21,6 +21,70 @@ use crate::{
     arch::boot::smp::get_num_processors, task::DisabledPreemptGuard, trap::DisabledLocalIrqGuard,
 };
 
+/// The ID of a CPU in the system.
+///
+/// If converting from/to an integer, the integer must start from 0 and be less
+/// than the number of CPUs.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct CpuId(u32);
+
+/// An error indicating that the CPU does not exist.
+#[derive(Debug)]
+pub struct NonexistentCPU(pub usize);
+
+impl CpuId {
+    /// Returns the CPU ID of the bootstrap processor (BSP).
+    pub const fn bsp() -> Self {
+        CpuId(0)
+    }
+
+    /// Converts the CPU ID to an `usize`.
+    pub const fn as_usize(self) -> usize {
+        self.0 as usize
+    }
+
+    /// Converts the CPU ID to an `u32`.
+    pub const fn as_u32(self) -> u32 {
+        self.0
+    }
+}
+
+impl TryFrom<usize> for CpuId {
+    type Error = NonexistentCPU;
+
+    fn try_from(value: usize) -> Result<Self, Self::Error> {
+        if value < num_cpus() {
+            Ok(CpuId(value as u32))
+        } else {
+            Err(NonexistentCPU(value))
+        }
+    }
+}
+
+impl TryFrom<u32> for CpuId {
+    type Error = NonexistentCPU;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        if (value as usize) < num_cpus() {
+            Ok(CpuId(value))
+        } else {
+            Err(NonexistentCPU(value as usize))
+        }
+    }
+}
+
+impl From<CpuId> for usize {
+    fn from(id: CpuId) -> usize {
+        id.0 as usize
+    }
+}
+
+impl From<CpuId> for u32 {
+    fn from(id: CpuId) -> u32 {
+        id.0
+    }
+}
```

Comment:
These methods seem to be redundant with `CpuId::as_u32` and `CpuId::as_usize`. Leaving only `From` trait methods should be enough.

Reply:
Um, no. `CpuId::as_u32` and `CpuId::as_usize` are constant functions that are allowed in constructors. There's no `const` version of `From` yet. So sadly they should be kept like this.